### PR TITLE
Python in production part 4 of 5

### PR DESCRIPTION
**Pull request by lward27**
This one is a bit long, but lot's of good information and lessons learned contained within.

### CODE REVIEW

**Comment by lward27** [comment]
From Nathan Shurte - switch unit file to user space, not system space. Seek slack thread https://ipponusa.slack.com/archives/C54D5S91B/p1666119730635589

**Review by lanefu — APPROVED** [review]
good job see suggestions

File: posts/python-in-production-part-4-of-5.md [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+
+[Service]
+ExecStart=/usr/local/bin/simple_api
+Restart=always
+
+[Install]
+WantedBy=multi-user.target
+```
+
+Following is an explanation of each line.
```
Comment (pkrish123) [inline posts/python-in-production-part-4-of-5.md:56]:
```suggestion
Here is an explanation of each line.
```

File: posts/python-in-production-part-4-of-5.md [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+
+Now that we have our systemd unit file, we are ready to put it, along with our program binary, onto our linux server.
+
+# Putting it on the Server
+
+**Note:** This section involves some pretty advanced Linux topics.  It's totally cool if you don't fully understand everything.  I will offer some minor explanations, but I highly encourage you to read up on any and all commands that you run on your machines.  Please don't *blindly run* linux commands that you find on the internet.
+
+If you have been following along since the beginning of the series, then you will have a *binary* file called `simple_api` and a *systemd unit file* called `simple_api.service` in your current working directory.  In order to get these files onto our linux server, we are going to use a utility called ***scp***.
+
+SCP stands for **S**ecure **C**opy.  It is a simple command line utility that transfers files from your local machine, to a remote server (or the other way around!).  If you are more comfortable with an ftp utility, or a program such as filezilla, feel free to use that instead.
```
Comment (pkrish123) [inline posts/python-in-production-part-4-of-5.md:74]: You've only highlighted the S & C of SCP. Isn't SCP, Secure Copy Protocol?

Comment (pkrish123) [reply posts/python-in-production-part-4-of-5.md:74]: You might also want to highlight what makes the transfer secure specifically here versus just running a cp

File: posts/python-in-production-part-4-of-5.md [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+
+### Copy the Binary File onto Your Linux Server
+
+If you are using an SSH key to access your server, then run this command on your local machine, replacing the stuff in brackets with *your stuff*:
+
+```bash
+scp -i {path_to_ssh_private_key_file} simple_api {user}@{server}:/home/{user}/simple_api
+```
+
+My command looks like this (DON'T RUN THIS ONE!):
```
Comment (pkrish123) [inline posts/python-in-production-part-4-of-5.md:84]:
```suggestion
My command looks like this (***DON'T RUN THIS ONE!***): 
```

Comment (pkrish123) [reply posts/python-in-production-part-4-of-5.md:84]: Bolding and italicizing for extra emphasis!

File: posts/python-in-production-part-4-of-5.md [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+```bash
+scp -i ~/.ssh/id_ed25519 simple_api lward@python-blog.example.com:/home/lward/simple_api
+```
+
+If you are accessing your server via username and password, simply run this command and follow the on screen prompts for your username and password:
+```bash
+scp simple_api {user}@{server}:/home/{user}/simple_api
+```
+
+An example (DON'T RUN THIS ONE!):
```
Comment (pkrish123) [inline posts/python-in-production-part-4-of-5.md:95]:
```suggestion
An example (***DON'T RUN THIS ONE!***):
```

File: posts/python-in-production-part-4-of-5.md [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+
+### Copy the SystemD Unit File onto Your Linux Server
+
+Again, run this command from your local machine, replacing the stuff in brackets with *your stuff*:
+
+```bash
+scp -i {path_to_ssh_private_key_file} simple_api.service {user}@{server}:/home/{user}/simple_api.service
+```
+
+The next step after copying both files on to our server, is to log into our server and complete our setup.  This next step will only need to be completed one time, and then it is good forever!
```
Comment (pkrish123) [inline posts/python-in-production-part-4-of-5.md:109]:
```suggestion
After copying both files onto our server, the next step is to complete our setup.  This next step will only need to be completed once; it is good forever!
```

File: posts/python-in-production-part-4-of-5.md [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+## Reload SystemD and Enable your Service
+
+Just having our files in place does not mean that systemd will just start running our application.  We have to reload the daemon to see our new unit file, and then enable and start our service.  Systemd includes `systemctl`, it is used to manage our system and user unit files.  While logged into your server, run these commands:
+```bash
+systemctl --user daemon-reload
+systemctl --user enable simple_api.service
+systemctl --user start simple_api.service
+```
+
+These three commands tell systemd to look for new unit files, add them to the list of things to start on boot, and then start the service.  To verify our service is indeed running, we will use another piece of software included with systemd, called `journald`.  In order to read the logs stored in `journald`, we use the command `journalctl`.  You may notice a patter emerging here!
```
Comment (pkrish123) [inline posts/python-in-production-part-4-of-5.md:144]: patter -> pattern

File: posts/python-in-production-part-4-of-5.md [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+bash build.sh
+```
+
+Let's test our application before sending it to the server.
+```bash
+export DEV_LOCATION="North Carolina"
+./simple_api
+```
+
+Use your browser to navigate to the site and you should see the new message `{"Message":"Hello, World! From North Carolina"}`.  When testing locally, we simply set an Environment Variable, but systemd actually has an option to load an environment file.  A few screens ago, we created a file called `simple_api.conf`.  We are going to put that file on our linux server in the `/etc/default/simple_api.conf` location.  Let's go ahead and update our unit file and then we will load it all onto our server.
```
Comment (pkrish123) [inline posts/python-in-production-part-4-of-5.md:231]:
```suggestion
Use your browser to navigate the site and see the new message `{"Message":"Hello, World! From North Carolina"}`.  When testing locally, we set an Environment Variable, but systemd has an option to load an environment file.  Earlier, we created a file called `simple_api.conf`.  We will put that file on our Linux server in the `/etc/default/simple_api.conf` location.  Let's go ahead and update our unit file and then we will load it all onto our server.
```

File: posts/python-in-production-part-4-of-5.md [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+```bash
+lward@python-blog:~$ curl http://localhost:8000
+{"Message":"Hello, World! From North Carolina"}
+```
+
+# Why You Should Use an Environment File vs Hard Coding
+
+If you have made it this far, then I salute you, dedicated reader.  This series consists of 5 parts, and you are nearing the end of part 4.  Each part builds upon the previous part, and each part offers more value as far as ***lessons learned***.  To illuminate this point, I will now tell you why you are creating this configuration file versus hard coding 'North Carolina' into the source code.
+
+Let's say that down the road, I move to Virgina.  If I had hardcoded my location into my source code, I would need to rebuild, and relaunch my appilication to reflect this change.  Using an environment file in our systemd unit file, enables us to change aspects of our program *without* having to rebuild it.  With a tiny application like this, rebuilding would not really be that big of a struggle, but things get complicated as they get bigger.  Let's update our location to Virgina and then call it a day.
```
Comment (pkrish123) [inline posts/python-in-production-part-4-of-5.md:278]:
```suggestion
Let's say that down the road, I move to Virginia.  If I had hardcoded my location into my source code, I would need to rebuild and relaunch my application to reflect this change.  Using an environment file in our systemd unit file enables us to change aspects of our program *without* having to rebuild it.  With a small application like this, rebuilding would be manageable, but things get complicated as applications get bigger.  Let's update our location to Virginia and then call it a day.
```

**Review by pkrish123 — CHANGES_REQUESTED** [review]

File: posts/python-in-production-part-4-of-5.md [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+tags:
+- python
+- api
+- devops
+- linux
+title: "Python in Production Part 4 of 5"
+image: 
+---
+
+Welcome back to the Python in Production series.  In the [third part](https://templinkplaceholder) of the series, we learned how to convert our program into a runnable binary file.  Now that we have a runnable binary, we are ready to *integrate* our program into a system.
```
Comment (pkrish123) [inline posts/python-in-production-part-4-of-5.md:13]: You can update the link for Part 3 now that it's published as well!

File: posts/python-in-production-part-4-of-5.md (hunk 4 above)
Comment (pkrish123) [reply posts/python-in-production-part-4-of-5.md:95]: Bolding and Italicizing for extra emphasis!

**Review by pkrish123 — APPROVED** [review]

**Pull request merged**